documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc

```diff
@@ -0,0 +1,232 @@
+++++
+
+
+In case of Local or Replicated mode, all data needs to fit in memory, so calculating the amount of required memory is trivial.
+
+Calculating memory requirements for Distributed mode is slightly more complicated and requires using the following:
+
+[stem]
+.Required memory in Distributed mode
+++++
+Required Memo ry = (Total Data Set)/(Nodes)*(1+(Node Failures)/(Nodes - Node Failures))
+++++
+
+Where:
+
+* `Total Data Set` - Estimated size of all data
+* `Nodes` - The number of nodes in the cluster
+* `Node Failures` - Number of possible failures (also `number of owners - 1`)
+
+Calculated amount of memory should be used for setting `Xmx` and `Xms` parameters.
+
+JVM as well as Infinispan require additional memory for other tasks like searches, allocating network buffers etc.
+It is advised to allocate no more than 50% of memory with living data.
+
+When considering large heaps, make sure there's enough CPU to perform garbage collection efficiently.
+
+== Java Virtual Machine settings
+
+Java Virtual Machine tuning might be divided into sections like memory or GC.
+Below is a list of helpful configuration parameters and a guide how to adjust them.
+
+=== Memory settings
+
+Adjusting memory size is one of the most crucial step in Infinispan tuning. The most commonly used JVM flags are:
+
+* `-Xms` - Defines the minimum heap size allowed.
+* `-Xmx` - Defines the maximum heap size allowed.
+* `-Xmn` - Defines the minimum and maximum value for the young generation.
+* `-XX:NewRatio` - Define the ratio between young and old generations. Should not be used if -Xmn is enabled.
+
+Using `Xms` equal to `Xmx` will prevent JVM from dynamically sizing memory and might decrease GC pauses cased by resizing.
+
+=== Garbage collection
+
+The main goal is to minimize the amount of time when JVM is paused. Having said that, CMS is a suggested GC for Infinispan applications.
+
+The most frequently used JVM flags are:
+
+* `XX:MaxGCPauseMillis` - Sets a target for the maximum GC pause time. Should be tuned to meet the SLA.
+* `-XX:+ UseConcMarkSweepGC` - Enables usage of the CMS collector.
+* `-XX:+ CMSClassUnloadingEnabled` - Allows class unloading when the CMS collector is enabled.
+* `-XX:+ UseParNewGC` - Utilize a parallel collector for the young generation. This parameter minimizes pausing by using multiple collection threads in parallel.
+* `-XX:+ DisableExplicitGC` - Prevent explicit garbage collections.
+
+=== Other settings
+
+There are two additional parameters which are suggested to be used:
+
+* `-server` - Enables server mode for the JVM.
+* `-XX:+ UseLargePages` - Instructs the JVM to allocate memory in Large Pages. These pages must be configured at the OS level for this parameter to function successfully.
+
+=== Example configuration
+
+.32GB JVM
+----
+-server
+-Xmx32G
+-Xms32G
+-Xmn8G
+-XX:+UseLargePages
+-XX:+UseConcMarkSweepGC
+-XX:+UseParNewGC
+-XX:+DisableExplicitGC
```

Comment:
Argh... It probably won't be that easy because if we squashed all Java sub-paragraphs (`Memory settings`, `Garbage collection` and `Other settings`) we will lose some background information like this:

> The main goal is to minimize the amount of time when JVM is paused. Having said that, CMS is a suggested GC for Infinispan applications.

or this:

> Using `Xms` equal to `Xmx` will prevent JVM from dynamically sizing memory and might decrease GC pauses caused by resizing. 
Also when using this combination, it is advices to specify the `Xmn` parameter as well. 

So the question is, do we still want to do it?

@alanfx - Would be great to hear your input here also.